File: Baekjoon/Gold 4/G4_2239_스도쿠.java
package a0327;

import java.util.*;
import java.io.*;

public class G4_2239_스도쿠 {
	static int[][] map = new int[9][9];
	static boolean[][] sudo_r = new boolean[10][10];
	static boolean[][] sudo_c = new boolean[10][10];
	static boolean[][][] sudo_3x3 = new boolean[3][3][10];
	static StringBuilder sb = new StringBuilder();
	
	public static void main(String[] args) throws Exception{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		
		for(int i=0; i<9; i++){
			String string = br.readLine();
			for(int j=0; j<9; j++) map[i][j] = string.charAt(j) - '0';
		}
		
		for(int i=0; i<9; i++){
			for(int j=0; j<9; j++){
				if(map[i][j] != 0){
					sudo_r[i][map[i][j]] = true;
					sudo_c[j][map[i][j]] = true;
					sudo_3x3[i/3][j/3][map[i][j]] = true;
				}
			}
		}
		DFS(0, 0);
		System.out.println(sb.toString());
	}
	
	static void DFS(int r, int c) {
		if(c==9) {
			c = 0;
			r++;
		}
		if(r==9){
			for(int i=0; i<9; i++){
				for(int j=0; j<9; j++) {
					sb.append(map[i][j]);
					System.out.print(map[i][j]);
				}
				System.out.println();
				sb.append("\n");
			}
			System.exit(0);
		}
		if(map[r][c]==0){
			for(int k=1; k<10; k++){
				if(!sudo_r[r][k] && !sudo_c[c][k] && !sudo_3x3[r/3][c/3][k]){
					sudo_r[r][k] = true;
					sudo_c[c][k] = true;
					sudo_3x3[r/3][c/3][k] = true;
					map[r][c] = k;
					map[r][c] = k;
					DFS(r, c+1);
					sudo_r[r][k] = false;
					sudo_c[c][k] = false;
					sudo_3x3[r/3][c/3][k] = false;
					map[r][c] = 0;
				}
			}
		}else{
			DFS(r, c+1);
		}
	}
}

File: Baekjoon/Gold4/G4_2239_스도쿠.py
# 각 행과 열을 모두 집합으로 만들어서 중복을 빠르게 검사할 수 있도록 한다.
# DFS로 백트래킹한다.
import sys; input=sys.stdin.readline

def DFS(G,sudo_r,sudo_c,sudo_3x3,r,c):
    if c==9:
        c=0
        r+=1
    if r==9:
        for i in range(9):
            for j in range(9):
                print(G[i][j],end='')
            print()
        exit()
    if not G[r][c]:
        for k in range(1,10):
            if k not in sudo_r[r] and k not in sudo_c[c] and k not in sudo_3x3[r//3][c//3]:
                sudo_r[r].add(k)
                sudo_c[c].add(k)
                sudo_3x3[r//3][c//3].add(k)
                G[r][c]=k
                DFS(G,sudo_r,sudo_c,sudo_3x3,r,c+1)
                sudo_r[r].discard(k)
                sudo_c[c].discard(k)
                sudo_3x3[r//3][c//3].discard(k)
                G[r][c]=0
    else:
        DFS(G,sudo_r,sudo_c,sudo_3x3,r,c+1)

def main():
    G=[list(input().rstrip()) for _ in range(9)]
    sudo_r=[set() for _ in range(9)]
    sudo_c=[set() for _ in range(9)]
    sudo_3x3=[[set() for i in range(3)] for j in range(3)]
    for i in range(9):
        for j in range(9):
            G[i][j]=int(G[i][j])
            if G[i][j]:
                sudo_r[i].add(G[i][j])
                sudo_c[j].add(G[i][j])
                sudo_3x3[i//3][j//3].add(G[i][j])
    DFS(G,sudo_r,sudo_c,sudo_3x3,0,0)

if __name__=='__main__':
    main()
